In [0]:
# Replace these with your actual credentials
ACCESS_KEY = "Access_key"
SECRET_KEY = "Secret_key"
REGION = "ap-southeast-2"  # or your actual region

In [0]:
# Configure S3 access
spark._jsc.hadoopConfiguration().set("fs.s3a.access.key", ACCESS_KEY)
spark._jsc.hadoopConfiguration().set("fs.s3a.secret.key", SECRET_KEY)
spark._jsc.hadoopConfiguration().set("fs.s3a.endpoint", f"s3.{REGION}.amazonaws.com")

In [0]:
csv_path = "s3a://dj-databricks-pipeline/owid-covid-data.csv"

df = spark.read.option("header", True).option("inferSchema", True).csv(csv_path)

display(df.show(5))


+--------+---------+-----------+----------+-----------+---------+------------------+------------+----------+-------------------+-----------------------+---------------------+------------------------------+------------------------+----------------------+-------------------------------+-----------------+------------+------------------------+-------------+-------------------------+---------------------+---------------------------------+----------------------+----------------------------------+---------+-----------+------------------------+----------------------+------------------+-------------------------------+-------------+--------------+-----------+------------------+-----------------+-----------------------+--------------+----------------+-------------------------+------------------------------+-----------------------------+-----------------------------------+--------------------------+-------------------------------------+------------------------------+-------------------------------

In [0]:
display(df.limit(5).toPandas()) #Since we are not able to read the data properly above now we will use toPandas functions to see the Data

iso_code continent location date total_cases new_cases new_cases_smoothed total_deaths new_deaths new_deaths_smoothed total_cases_per_million new_cases_per_million new_cases_smoothed_per_million total_deaths_per_million new_deaths_per_million new_deaths_smoothed_per_million reproduction_rate icu_patients icu_patients_per_million hosp_patients hosp_patients_per_million weekly_icu_admissions weekly_icu_admissions_per_million weekly_hosp_admissions weekly_hosp_admissions_per_million new_tests total_tests total_tests_per_thousand new_tests_per_thousand new_tests_smoothed new_tests_smoothed_per_thousand positive_rate tests_per_case tests_units total_vaccinations people_vaccinated people_fully_vaccinated total_boosters new_vaccinations new_vaccinations_smoothed total_vaccinations_per_hundred people_vaccinated_per_hundred people_fully_vaccinated_per_hundred total_boosters_per_hundred new_vaccinations_smoothed_per_million new_people_vaccinated_smoothed new_people_vaccinated_smoothed_per_hundred stringency_index population population_density median_age aged_65_older aged_70_older gdp_per_capita extreme_poverty cardiovasc_death_rate diabetes_prevalence female_smokers male_smokers handwashing_facilities hospital_beds_per_thousand life_expectancy human_development_index excess_mortality_cumulative_absolute excess_mortality_cumulative excess_mortality excess_mortality_cumulative_per_million AFG Asia Afghanistan 2020-02-24 5.0 5.0 null null null null 0.126 0.126 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-25 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-26 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-27 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-28 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null

In [0]:
df.select("iso_code").distinct().count() #Let us check the count of all unique values present in the iso_code

Out[5]: 238

In [0]:
df.groupBy("iso_code").count().show() #TWe will be seeing the count of each country.

+--------+-----+
|iso_code|count|
+--------+-----+
|     BRB|  719|
|     BRA|  739|
|     ARM|  762|
|     ABW|  723|
|     BRN|  727|
|     BOL|  725|
|     ATG|  723|
|OWID_AFR|  752|
|     BHS|  720|
|     BMU|  717|
|     AZE|  735|
|     AUS|  770|
|     AFG|  741|
|     BGD|  733|
|     BLR|  737|
|     BWA|  706|
|     ALB|  740|
|     BHR|  741|
|     BIH|  731|
|     AND|  734|
+--------+-----+
only showing top 20 rows



In [0]:
#LEt us see the count of the continents

df.select("continent").distinct().count()

Out[7]: 7

We do have the data for all the 7 continents.

In [0]:
#Now let us see the count of each continents - how many values are there for each of them.

df.groupBy("continent").count().show()

+-------------+-----+
|    continent|count|
+-------------+-----+
|       Europe|37088|
|       Africa|39417|
|         null| 9956|
|North America|24931|
|South America| 9517|
|      Oceania| 9217|
|         Asia|36200|
+-------------+-----+



In [0]:
#Importing the important libraries
from pyspark.sql.functions import col


#Let us try to change the name of column to something appropriate and keep them sorted by contient name
df.groupBy("continent")\
    .count()\
    .withColumnRenamed("count","countOfRecords")\
    .orderBy(col("continent"))\
    .show()

+-------------+--------------+
|    continent|countOfRecords|
+-------------+--------------+
|         null|          9956|
|       Africa|         39417|
|         Asia|         36200|
|       Europe|         37088|
|North America|         24931|
|      Oceania|          9217|
|South America|          9517|
+-------------+--------------+



This shows that most of the records belong to Africa, Asia and Europe with count of more than 30000+

In [0]:
#Let us check what are the values of Countries associated with continents where we have null values

count_of_values = df.filter(col("continent").isNull())\
    .select("location").distinct().count()

print(f"There are {count_of_values} unique values associated with Null values of continent.")

df.filter(col("continent").isNull())\
    .groupBy("location").count().withColumnRenamed("count","countOfRecords")\
        .orderBy("countOfRecords",ascending=False).show()

There are 13 unique values associated with Null values of continent.
+-------------------+--------------+
|           location|countOfRecords|
+-------------------+--------------+
|               Asia|           774|
|        High income|           774|
|Lower middle income|           774|
|      North America|           774|
|              World|           774|
|Upper middle income|           774|
|             Europe|           773|
|     European Union|           773|
|            Oceania|           771|
|      International|           758|
|             Africa|           752|
|      South America|           743|
|         Low income|           742|
+-------------------+--------------+



Here we can see that count for all the 13 unique values is almost similar and they stand close to 740. 

In [0]:
## We can remove the rows where continent is having NULL values because the value of location does not seem to be having any sense
# for example we have values like High Income, Upper middle income etc.

df_filtered_1 = df.filter(col("continent").isNotNull())



In [0]:
display(df_filtered_1.limit(5))

iso_code continent location date total_cases new_cases new_cases_smoothed total_deaths new_deaths new_deaths_smoothed total_cases_per_million new_cases_per_million new_cases_smoothed_per_million total_deaths_per_million new_deaths_per_million new_deaths_smoothed_per_million reproduction_rate icu_patients icu_patients_per_million hosp_patients hosp_patients_per_million weekly_icu_admissions weekly_icu_admissions_per_million weekly_hosp_admissions weekly_hosp_admissions_per_million new_tests total_tests total_tests_per_thousand new_tests_per_thousand new_tests_smoothed new_tests_smoothed_per_thousand positive_rate tests_per_case tests_units total_vaccinations people_vaccinated people_fully_vaccinated total_boosters new_vaccinations new_vaccinations_smoothed total_vaccinations_per_hundred people_vaccinated_per_hundred people_fully_vaccinated_per_hundred total_boosters_per_hundred new_vaccinations_smoothed_per_million new_people_vaccinated_smoothed new_people_vaccinated_smoothed_per_hundred stringency_index population population_density median_age aged_65_older aged_70_older gdp_per_capita extreme_poverty cardiovasc_death_rate diabetes_prevalence female_smokers male_smokers handwashing_facilities hospital_beds_per_thousand life_expectancy human_development_index excess_mortality_cumulative_absolute excess_mortality_cumulative excess_mortality excess_mortality_cumulative_per_million AFG Asia Afghanistan 2020-02-24 5.0 5.0 null null null null 0.126 0.126 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-25 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-26 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-27 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-28 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null

In [0]:
## Let us check few rows for a country - India


display(df_filtered_1.filter(col("location")=="India").limit(5))

iso_code continent location date total_cases new_cases new_cases_smoothed total_deaths new_deaths new_deaths_smoothed total_cases_per_million new_cases_per_million new_cases_smoothed_per_million total_deaths_per_million new_deaths_per_million new_deaths_smoothed_per_million reproduction_rate icu_patients icu_patients_per_million hosp_patients hosp_patients_per_million weekly_icu_admissions weekly_icu_admissions_per_million weekly_hosp_admissions weekly_hosp_admissions_per_million new_tests total_tests total_tests_per_thousand new_tests_per_thousand new_tests_smoothed new_tests_smoothed_per_thousand positive_rate tests_per_case tests_units total_vaccinations people_vaccinated people_fully_vaccinated total_boosters new_vaccinations new_vaccinations_smoothed total_vaccinations_per_hundred people_vaccinated_per_hundred people_fully_vaccinated_per_hundred total_boosters_per_hundred new_vaccinations_smoothed_per_million new_people_vaccinated_smoothed new_people_vaccinated_smoothed_per_hundred stringency_index population population_density median_age aged_65_older aged_70_older gdp_per_capita extreme_poverty cardiovasc_death_rate diabetes_prevalence female_smokers male_smokers handwashing_facilities hospital_beds_per_thousand life_expectancy human_development_index excess_mortality_cumulative_absolute excess_mortality_cumulative excess_mortality excess_mortality_cumulative_per_million IND Asia India 2020-01-30 1.0 1.0 null null null null 0.001 0.001 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 10.19 1.393409033E9 450.419 28.2 5.989 3.414 6426.674 21.2 282.28 10.39 1.9 20.6 59.55 0.53 69.66 0.645 null null null null IND Asia India 2020-01-31 1.0 0.0 null null null null 0.001 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 10.19 1.393409033E9 450.419 28.2 5.989 3.414 6426.674 21.2 282.28 10.39 1.9 20.6 59.55 0.53 69.66 0.645 null null null null IND Asia India 2020-02-01 1.0 0.0 null null null null 0.001 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 10.19 1.393409033E9 450.419 28.2 5.989 3.414 6426.674 21.2 282.28 10.39 1.9 20.6 59.55 0.53 69.66 0.645 null null null null IND Asia India 2020-02-02 2.0 1.0 null null null null 0.001 0.001 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 10.19 1.393409033E9 450.419 28.2 5.989 3.414 6426.674 21.2 282.28 10.39 1.9 20.6 59.55 0.53 69.66 0.645 null null null null IND Asia India 2020-02-03 3.0 1.0 null null null null 0.002 0.001 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 10.19 1.393409033E9 450.419 28.2 5.989 3.414 6426.674 21.2 282.28 10.39 1.9 20.6 59.55 0.53 69.66 0.645 null null null null

In [0]:
#let us check the count of distinct values of population for india

unique_population_values = df_filtered_1.filter(col("location")=="India")\
    .select("population").distinct().count()

print(f"Total unique values of population for India is {unique_population_values}")

df_filtered_1.filter(col("location")=="India")\
    .select("population").distinct().show()

Total unique values of population for India is 1
+-------------+
|   population|
+-------------+
|1.393409033E9|
+-------------+



In [0]:
#Let us see for each country how many unique values of population they have

population_values_df = df_filtered_1.select("location","population")\
                                                    .distinct().groupBy("location").count().toPandas()



# #Let us check the location where population is greater than 1
population_values_df[population_values_df['count']>1]

location 
 count

This tells us that we do not have any location where population is varying, it is having same value for each country throughout the dataset. So we can remove the column as it is constant for each country. 

In [0]:
## We can create a dataframe where we can keep the avg and median value of the cases for a country, and avg median values of
# new cases each day.


display(df_filtered_1.limit(5).toPandas())

iso_code continent location date total_cases new_cases new_cases_smoothed total_deaths new_deaths new_deaths_smoothed total_cases_per_million new_cases_per_million new_cases_smoothed_per_million total_deaths_per_million new_deaths_per_million new_deaths_smoothed_per_million reproduction_rate icu_patients icu_patients_per_million hosp_patients hosp_patients_per_million weekly_icu_admissions weekly_icu_admissions_per_million weekly_hosp_admissions weekly_hosp_admissions_per_million new_tests total_tests total_tests_per_thousand new_tests_per_thousand new_tests_smoothed new_tests_smoothed_per_thousand positive_rate tests_per_case tests_units total_vaccinations people_vaccinated people_fully_vaccinated total_boosters new_vaccinations new_vaccinations_smoothed total_vaccinations_per_hundred people_vaccinated_per_hundred people_fully_vaccinated_per_hundred total_boosters_per_hundred new_vaccinations_smoothed_per_million new_people_vaccinated_smoothed new_people_vaccinated_smoothed_per_hundred stringency_index population population_density median_age aged_65_older aged_70_older gdp_per_capita extreme_poverty cardiovasc_death_rate diabetes_prevalence female_smokers male_smokers handwashing_facilities hospital_beds_per_thousand life_expectancy human_development_index excess_mortality_cumulative_absolute excess_mortality_cumulative excess_mortality excess_mortality_cumulative_per_million AFG Asia Afghanistan 2020-02-24 5.0 5.0 null null null null 0.126 0.126 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-25 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-26 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-27 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null AFG Asia Afghanistan 2020-02-28 5.0 0.0 null null null null 0.126 0.0 null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null null 8.33 3.9835428E7 54.422 18.6 2.581 1.337 1803.987 null 597.029 9.59 null null 37.746 0.5 64.83 0.511 null null null null

In [0]:
## We can create three analysis majorly now -


# 1. What is the average and median number of cases we are getting each day overall.
# 2. What is the average and median number of cases we are getting each country - new cases
# 3. What is the average and median number of cases we are getting each country - total cases


In [0]:
df_filtered_1.select("date").printSchema()

root
 |-- date: date (nullable = true)



In [0]:
from pyspark.sql.functions import avg, expr, sum



## Let us get the average and median of new cases and total cases 

new_cases = df_filtered_1.groupBy("date").agg(
    avg("new_cases").alias("average_new_Cases"),
    expr("percentile_approx(new_cases, 0.5)").alias("medain_new_cases"),
    sum("new_cases").alias("sum_of_new_cases")
).orderBy("date")



new_cases = new_cases.filter(col("date")>"2020-01-22")

new_cases.show()

+----------+------------------+----------------+----------------+
|      date| average_new_Cases|medain_new_cases|sum_of_new_cases|
+----------+------------------+----------------+----------------+
|2020-01-23| 9.090909090909092|             1.0|           100.0|
|2020-01-24|23.916666666666668|             1.0|           287.0|
|2020-01-25|35.214285714285715|             0.0|           493.0|
|2020-01-26| 45.53333333333333|             1.0|           683.0|
|2020-01-27| 44.94444444444444|             0.0|           809.0|
|2020-01-28|147.27777777777777|             0.0|          2651.0|
|2020-01-29|             29.45|             0.0|           589.0|
|2020-01-30|              94.0|             0.0|          2068.0|
|2020-01-31|              67.6|             0.0|          1690.0|
|2020-02-01| 78.18518518518519|             0.0|          2111.0|
|2020-02-02|175.88888888888889|             0.0|          4749.0|
|2020-02-03|114.81481481481481|             0.0|          3100.0|
|2020-02-0

After further analysis we got to know that there are nUll values for the new cases till 22-Jan-2020. We will filter that data from our new dataset. 

In [0]:
## Let us see same thing for total cases



total_cases = df_filtered_1.groupBy("date").agg(
    avg("total_cases").alias("average_total_Cases"),
    expr("percentile_approx(total_cases, 0.5)").alias("medain_total_cases"),
    sum("total_cases").alias("sum_of_total_cases")
).orderBy("date")


total_cases = total_cases.filter(col("date")>"2020-01-22")


total_cases.show()

+----------+-------------------+------------------+------------------+
|      date|average_total_Cases|medain_total_cases|sum_of_total_cases|
+----------+-------------------+------------------+------------------+
|2020-01-23|  59.72727272727273|               2.0|             657.0|
|2020-01-24|  78.66666666666667|               2.0|             944.0|
|2020-01-25| 102.64285714285714|               3.0|            1437.0|
|2020-01-26| 141.33333333333334|               4.0|            2120.0|
|2020-01-27| 162.72222222222223|               4.0|            2929.0|
|2020-01-28|              310.0|               4.0|            5580.0|
|2020-01-29|             308.45|               5.0|            6169.0|
|2020-01-30| 374.40909090909093|               4.0|            8237.0|
|2020-01-31|             397.08|               5.0|            9927.0|
|2020-02-01| 445.85185185185185|               6.0|           12038.0|
|2020-02-02|  621.7407407407408|               6.0|           16787.0|
|2020-

In [0]:
## Let us check the same thing for total_deaths


total_deaths = df_filtered_1.groupBy("date").agg(
    avg("total_deaths").alias("average_total_deaths"),
    expr("percentile_approx(total_deaths, 0.5)").alias("medain_total_death"),
    sum("total_deaths").alias("sum_of_total_deaths")
).orderBy("date")


total_deaths = total_deaths.filter(col("date")>"2020-01-22")


total_deaths.show()

+----------+--------------------+------------------+-------------------+
|      date|average_total_deaths|medain_total_death|sum_of_total_deaths|
+----------+--------------------+------------------+-------------------+
|2020-01-23|                18.0|              18.0|               18.0|
|2020-01-24|                26.0|              26.0|               26.0|
|2020-01-25|                42.0|              42.0|               42.0|
|2020-01-26|                56.0|              56.0|               56.0|
|2020-01-27|                82.0|              82.0|               82.0|
|2020-01-28|               131.0|             131.0|              131.0|
|2020-01-29|               133.0|             133.0|              133.0|
|2020-01-30|               171.0|             171.0|              171.0|
|2020-01-31|               213.0|             213.0|              213.0|
|2020-02-01|               259.0|             259.0|              259.0|
|2020-02-02|               181.0|               1.0

In [0]:
## We can merge this dataframe into one and store them into s3 as parquet

covid_Cases_data_day_wise = new_cases.join(total_cases, on="date", how="left")

covid_Cases_and_death_day_wise = covid_Cases_data_day_wise.join(total_deaths, on="date", how="left")

covid_Cases_and_death_day_wise.show()


+----------+------------------+----------------+----------------+-------------------+------------------+------------------+--------------------+------------------+-------------------+
|      date| average_new_Cases|medain_new_cases|sum_of_new_cases|average_total_Cases|medain_total_cases|sum_of_total_cases|average_total_deaths|medain_total_death|sum_of_total_deaths|
+----------+------------------+----------------+----------------+-------------------+------------------+------------------+--------------------+------------------+-------------------+
|2020-01-23| 9.090909090909092|             1.0|           100.0|  59.72727272727273|               2.0|             657.0|                18.0|              18.0|               18.0|
|2020-01-24|23.916666666666668|             1.0|           287.0|  78.66666666666667|               2.0|             944.0|                26.0|              26.0|               26.0|
|2020-01-25|35.214285714285715|             0.0|           493.0| 102.6428571428

In [0]:
covid_Cases_and_death_day_wise.write.mode("overwrite") \
    .parquet("s3a://dj-databricks-pipeline/day_wise_parquet/")

## Let us do same thing wrt country

In [0]:

## Let us get the average and median of new cases and total cases  related to country

new_cases_country_wise = df_filtered_1.groupBy("location").agg(
    avg("new_cases").alias("average_new_Cases"),
    expr("percentile_approx(new_cases, 0.5)").alias("medain_new_cases"),
    sum("new_cases").alias("sum_of_new_cases")
).orderBy("location")


new_cases_country_wise.show()

+-------------------+------------------+----------------+----------------+
|           location| average_new_Cases|medain_new_cases|sum_of_new_cases|
+-------------------+------------------+----------------+----------------+
|        Afghanistan|235.91204330175913|            78.0|        174339.0|
|            Albania| 374.6969696969697|           182.0|        272030.0|
|            Algeria|358.41486486486485|           235.0|        265227.0|
|            Andorra| 52.36239782016349|            12.0|         38434.0|
|             Angola| 137.9832402234637|            71.0|         98796.0|
|           Anguilla|3.6087570621468927|             0.0|          2555.0|
|Antigua and Barbuda| 10.34257975034674|             0.0|          7457.0|
|          Argentina|12188.714870395634|          7208.0|       8934328.0|
|            Armenia| 573.0965986394558|           339.0|        421226.0|
|              Aruba|46.655601659751035|            17.0|         33732.0|
|          Australia| 441

In [0]:

toal_cases_country_wise = df_filtered_1.groupBy("location").agg(
    avg("total_cases").alias("average_total_Cases"),
    expr("percentile_approx(total_cases, 0.5)").alias("medain_total_cases")
).orderBy("location")


toal_cases_country_wise.show()

+-------------------+-------------------+------------------+
|           location|average_total_Cases|medain_total_cases|
+-------------------+-------------------+------------------+
|        Afghanistan|  79066.79217273954|           55714.0|
|            Albania|   97527.1636863824|          112897.0|
|            Algeria| 112892.52162162162|          113092.0|
|            Andorra| 10809.235694822888|           10948.0|
|             Angola| 31095.916201117318|           21265.0|
|           Anguilla| 391.63700564971754|              21.0|
|Antigua and Barbuda| 1548.6874135546334|             862.0|
|          Argentina|  2861118.507503411|         2133963.0|
|            Armenia| 170162.47755102042|          172816.0|
|              Aruba|  9608.976486860305|            8177.0|
|          Australia| 238310.98571428572|           28898.0|
|            Austria|  537652.1702702703|          459440.0|
|         Azerbaijan|  264725.0136054422|          235333.0|
|            Bahamas|  1

In [0]:
#Joining both dataframes
country_wise_cases = new_cases_country_wise.join(toal_cases_country_wise, on="location", how="left")

country_wise_cases.show()

+-------------------+------------------+----------------+----------------+-------------------+------------------+
|           location| average_new_Cases|medain_new_cases|sum_of_new_cases|average_total_Cases|medain_total_cases|
+-------------------+------------------+----------------+----------------+-------------------+------------------+
|        Afghanistan|235.91204330175913|            78.0|        174339.0|  79066.79217273954|           55714.0|
|            Albania| 374.6969696969697|           182.0|        272030.0|   97527.1636863824|          112897.0|
|            Algeria|358.41486486486485|           235.0|        265227.0| 112892.52162162162|          113092.0|
|            Andorra| 52.36239782016349|            12.0|         38434.0| 10809.235694822888|           10948.0|
|             Angola| 137.9832402234637|            71.0|         98796.0| 31095.916201117318|           21265.0|
|           Anguilla|3.6087570621468927|             0.0|          2555.0| 391.637005649

In [0]:
country_wise_cases.write.mode("overwrite") \
    .parquet("s3a://dj-databricks-pipeline/country_wise_parquet/")